File: LeetCode/22. Generate Parentheses/elixir/solution.exs
defmodule Solution do
  @spec generate_parenthesis(n :: integer) :: [String.t()]
  def generate_parenthesis(n) do
    n = n * 2

    0..trunc(2 ** n - 1)
    |> Enum.map(fn x ->
      Integer.to_string(x, 2)
      |> String.replace(["0", "1"], fn
        "0" -> "("
        "1" -> ")"
      end)
      |> String.pad_leading(n, "(")
    end)
    |> Enum.filter(&is_valid/1)
  end

  defp is_valid(s) do
    s
    |> to_charlist()
    |> valid_parentheses()
  end

  defp valid_parentheses(chars, stack \\ [])

  defp valid_parentheses([], stack) do
    stack == []
  end

  defp valid_parentheses([h | t], []) do
    valid_parentheses(t, [h])
  end

  defp valid_parentheses([h | t], [top | rest] = stack) do
    if is_complete_brackets(h, top) do
      valid_parentheses(t, rest)
    else
      valid_parentheses(t, [h | stack])
    end
  end

  defp is_complete_brackets(a, b) do
    case {b, a} do
      {?(, ?)} -> true
      {?{, ?}} -> true
      {?[, ?]} -> true
      _ -> false
    end
  end
end

File: LeetCode/22. Generate Parentheses/java/Solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> answer = new ArrayList<>();
        int length = 2 * n;
        int[] brackets;
        for (int i = 0; i < Math.pow(2, length); i++) {
            brackets = makeBracket(i, length);
            if (isValid(brackets)) {
                String bracketStr = bracketToString(brackets);
                answer.add(bracketStr);
            }
        }
        return answer;
    }

    private int[] makeBracket(int i, int length) {
        int[] brackets = new int[length];
        Arrays.fill(brackets, 1);
        char[] chars = Integer.toBinaryString(i).toCharArray();
        int charLength = chars.length;
        for (int j = 0; j < charLength; j++) {
            brackets[length - charLength + j] = chars[j] == '0' ? 1 : -1;
        }
        return brackets;
    }

    private String bracketToString(int[] brackets) {
        StringBuilder sb = new StringBuilder();
        for (int b : brackets) {
            sb.append(b == 1 ? "(" : ")");
        }
        return sb.toString();
    }

    private boolean isValid(int[] brackets) {
        int sum = 0;
        for (int i : brackets) {
            if (sum < 0) {
                return false;
            }
            sum += i;
        }
        return sum == 0;
    }
}